xbmc/interfaces/legacy/Addon.h

```diff
@@ -100,10 +100,155 @@ namespace XBMCAddon
 #ifdef DOXYGEN_SHOULD_USE_THIS
       ///
       /// \ingroup python_xbmcaddon
-      /// @brief \python_func{ xbmcaddon.Addon([id]).getSettings() }
+      /// @brief \python_func{ xbmcaddon.Addon([id]).supportsInstanceSettings() }
+      /// Returns true if the requested add-on supports several instances.
+      ///
+      /// @return Returns true if supported, false otherwise
+      ///
+      ///
+      ///-----------------------------------------------------------------------
+      /// @python_v21 New function added.
+      ///
+      ///
+      /// **Example:**
+      /// ~~~~~~~~~~~~~{.py}
+      /// ..
+      /// supportsInstances = self.Addon('pvr.iptvsimple').supportsInstanceSettings()
+      /// ..
+      /// ~~~~~~~~~~~~~
+      ///
+      supportsInstanceSettings();
+#else
+      bool supportsInstanceSettings();
+#endif
+
+#ifdef DOXYGEN_SHOULD_USE_THIS
+      ///
+      /// \ingroup python_xbmcaddon
+      /// @brief \python_func{ xbmcaddon.Addon([id]).getInstanceIds() }
+      /// Returns the list of known available instance ids.
+      ///
+      /// @return List of used add-on instances
+      ///
+      ///
+      ///-----------------------------------------------------------------------
+      /// @python_v21 New function added.
+      ///
+      ///
+      /// **Example:**
+      /// ~~~~~~~~~~~~~{.py}
+      /// ..
+      /// usedInstances = self.Addon('pvr.iptvsimple').getInstanceIds()
+      /// ..
+      /// ~~~~~~~~~~~~~
+      ///
+      getInstanceIds();
+#else
+      std::vector<unsigned int> getInstanceIds();
+#endif
+
+#ifdef DOXYGEN_SHOULD_USE_THIS
+      ///
+      /// \ingroup python_xbmcaddon
+      /// @brief \python_func{ xbmcaddon.Addon([id]).getFreeNewInstanceId() }
+      /// Get free add-on instance id to set new independent instance and settings about.
+      ///
+      /// @note After setting the new instance settings a @ref setInstanceState() call is required.
+      ///
+      /// @return Next usable and free instance id
+      ///
+      ///
+      ///-----------------------------------------------------------------------
+      /// @python_v21 New function added.
+      ///
+      ///
+      /// **Example:**
+      /// ~~~~~~~~~~~~~{.py}
+      /// ..
+      /// addon = xbmcaddon.Addon('pvr.iptvsimple')
+      /// newinstanceid = addon.getFreeNewInstanceId()
+      /// ..
+      /// ~~~~~~~~~~~~~
+      ///
+      getFreeNewInstanceId();
+#else
+      unsigned int getFreeNewInstanceId();
+#endif
+
+#ifdef DOXYGEN_SHOULD_USE_THIS
+      ///
+      /// \ingroup python_xbmcaddon
+      /// @brief \python_func{ xbmcaddon.Addon([id]).setInstanceState(id, enabled, [name]) }
```

Comment:
Why is the name set when change the state on the instance and not right when the instance is created? Can it be moved to getFreeNewInstanceId()?

Reply:
There is a big hen and egg problem, by `getFreeNewInstanceId()` is and can the instance not created, as the needed settings for it not available and PVR system then start works where always ends in errors, thats why it only give a free id. 

The fields about "enabled" and "name" are parts only handled by Kodi itself, about "enabled" is easy to understand why, about name relates to the name used on the GUI in Kodi (see picture above).

The `setInstanceState` needs be called on new added instances, thats why name needed on it, or can be called too on already present ones to change name and/or enable / disable it, thats why name declared optional.

Maybe the function can become another name. Have you a idea about?

----------------------

Another solution can be, but then it becomes hacky and needs more functions and much changes to add:
| Function | Hacky description |
|--|--|
| `settings = addon.createInstanceSettings(enabled, name)` | It returns the settings class and not an id<br> - The enabled and name needs temporary stored somewhere<br> - access to `CSettingsBase` in Kodi no more enough and needs access to `CAddonSettings` as this handle the id.
| `settings.setAddedInstanceReady()`  or<br> `addon.setAddedInstanceReady(settings)`| Something needs always done to report Kodi that new instance is ready to start and thats PVR system does it<br><br>Maybe it can optionally also be done by the settings class destructor, but then Python needs to do destruct it and never stores it somewhere.
| `id = settings.getInstanceId()` | New function then needed for the cases if python needs the id byself to store something about for him himself or for the future works.
| `addon.setInstanceState(id, enabled)` | To allow on already present ones the enable / disable
| `addon.setInstanceName(id, name)` | To allow on already present ones the name change
| `state = settings.getInstanceState()` and<br>`state = addon.getInstanceState(id)` | For the sake of form, to have the counterpart to the “set”.
| `name = settings.getInstanceName()` and<br>`name = addon.getInstanceName(id)` | For the sake of form, to have the counterpart to the “set”.

About this it seems then also more hard to understand what does somewhere what.